### ISSUE TITLE
[Sync] Update locally created records with server ids

### ISSUE THREAD

**Issue by davideweaver** [Nozbe/WatermelonDB#216]
Using PushChanges with a locally created record uses the `id` generated on the client. My server ignores this `id` and generates it's own `id`. What's the recommended way to update the local record with the `id` returned from the server? If I don't update the local `id` of the record, it will never receive updates from the server. Thanks

**Comment by radex** [comment]
@davideweaver is there no way for your server to just accept client's ID?

**Comment by davideweaver** [comment]
No. It’s a micro-service backend and I do not control all the services that I’ll need to sync/create records in.

On Jan 25, 2019, at 4:17 AM, Radek Pietruszewski <notifications@github.com<mailto:notifications@github.com>> wrote:


@davideweaver<https://github.com/davideweaver> is there no way for your server to just accept client's ID?

**Comment by esbenp** [comment]
In our API (we have sequential IDs) we have DB table that maps all database IDs to tablet IDs. So everytime we send data to the tablet we replace all internal IDs with the right tablet IDs. We also replace IDs of related entities etc. We also use this table to track deletes (so if you delete something in our DB we set a deleted_at timestamp on the tablet_id table).

**Comment by davideweaver** [comment]
@esbenp Did you create the tablet id mapping mechanism to work around the fact that there's no way to update ids for records created on your tablets? With zero knowledge of your stack, it seems like it would be easier to make the ids correct on the tablets.

**Comment by davideweaver** [comment]
@radex Any additional thoughts on this?

**Comment by davideweaver** [comment]
@radex Following up on the previous code, it fails because the update path in the adapter does not allow updates to the `id` as it's currently implemented. `encodeUpdate` uses the updated `id` to identify the record to be updated, so if `id` is changed the sql command doesn't find the record to update. Makes sense.

So I can fork this repo and get it to work for my purposes. Ideally this change could make it's way into your repo, if it's something you want. Let me know if that's something you'd like to talk about.

**Comment by radex** [comment]
> So I can fork this repo and get it to work for my purposes. Ideally this change could make it's way into your repo, if it's something you want. Let me know if that's something you'd like to talk about.

@davideweaver As I think you can understand, I'm hesitant towards accepting such changes into Watermelon. The problem is that "the ID of a record never changes" is an assumption that's probably hardcoded into a lot of places, and breaking this assumption could cause a lot of Watermelon stuff to break (perhaps in non-trivial ways). If you want to make a fork and work on this, I very much encourage you to do this — and if it turns out it's not so bad, I'll be very happy to pull your changes 🎉 

But perhaps this isn't the best way to approach it? @fletling and his team, for example, chose to use a system to convert between local and remote IDs during synchronization. This has its downsides too, but this way, the trouble with local and remote IDs is only concentrated on the interface between local and remote database. So it's easier to figure out all ways it could go wrong and protect against them — and requires no (or few) changes to Watermelon.

Of course, we think it's best to have consistent IDs between local and remote databases, but understandably, this might not always be possible to people — and you probably noticed a lot of similar issues dealing with ID issues. So it would be great to figure out a Watermelon best practice for this.

**Comment by davideweaver** [comment]
@radex Thanks for the follow-up. I understand that what I'm trying to do may not fit with your vision. At the same time, I am unable to control the ids of the service I'm syncing with and don't especially like the idea of maintaining a map between local and remote ids. So here's how I managed to make it work for now...

In `markLocalChangesAsSynced` I replace the record having the local id with a cloned record using the remote id. I'm sure there's a better way to clone the original record, but it seems to be working. The components observing the collection are behaving correctly as well.

```javascript
export default function markLocalChangesAsSynced(
  db: Database,
  syncedLocalChanges: SyncLocalChanges,
  pushResults: ?SyncPushResult
): Promise<void> {
  ensureActionsEnabled(db)
  return db.action(async () => {
    // update and destroy records concurrently
    await Promise.all([
      db.batch(...map(prepareMarkAsSynced, recordsToMarkAsSynced(syncedLocalChanges))),
      destroyDeletedRecords(db, syncedLocalChanges),
    ])

    // pushResults is an optional return value from pushChanges
    // that contains the mapping from remote to local ids per collection
    if (pushResults && pushResults.updatedCreateTables) {
      for (const tableKey: TableName<any> in pushResults.updatedCreateTables) {
        if (tableKey) {
          const table = pushResults.updatedCreateTables[tableKey]
          for (const id in table) {
            const collection = db.collections.get(tableKey)
            const model = await collection.find(id)
            if (model) {
              await collection._destroyPermanently(model)
              await collection.create(record => {
                const raw = {}
                for (const key in model._raw) {
                  if (key && key !== 'id' && key[0] !== '_') {
                    raw[key] = model._raw[key]
                  }
                }
                for (const key in model) {
                  if (key && key !== 'id' && key !== 'collection' && key !== 'syncStatus' && key[0] !== '_') {
                    record[key] = model[key]
                  }
                }
                raw.id = table[id]
                raw._status = 'synced'
                record._raw = sanitizedRaw(raw, collection.schema)
              })
            }
          }
        }
      }
    }

  }, 'sync-markLocalChangesAsSynced')
}
```

**Comment by cesargdm** [comment]
I would like to set my custom MongoDB IDs as the ids in the local DB so i can query easier a object based on the MongoDB ID I already retrieve by my backend service.

**Comment by Bessonov** [comment]
Yea, still relevant.

**Comment by Skullcan** [comment]
> I would like to set my custom MongoDB IDs as the ids in the local DB so i can query easier a object based on the MongoDB ID I already retrieve by my backend service.
> 
> [UPDATE] Found a way at #7

Find this thread, trying to achiev same you did. Lookd the #7 thread but haven't found how you change the ID. Can someone help me plz?

**Comment by diegolmello** [comment]
Take a look at this file on our repo: https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/develop/app/lib/methods/getRoles.js

It's a simple insert/update function (no delete) of roles requested from the server.
Note that we fetch all roles from watermelon and compare against the response to create or update.
It's verbose, but it works.

cc @radex just to check if there's a better idea on this case 😛

**Comment by Skullcan** [comment]
@diegolmello thank you! That´s what I was looking for.

**Comment by Bessonov** [comment]
Yeah, still relevant.

**Comment by whalemare** [comment]
I have my own backend, that IDs I can't track. 
Why you prefer client side ids instead backend, if data that you get, received from backend 🧐

**Comment by radex** [comment]
> Why you prefer client side ids instead backend

No, we prefer client and backend IDs being equivalent, because it makes things a lot simpler. Explained here: https://github.com/Nozbe/WatermelonDB/blob/master/docs-master/Advanced/Sync.md#local-vs-remote-ids

**Comment by KennanChan** [comment]
In my project I did some trick on this to solve the id problem

The client side database generates its own random id for a local record "A".
The backend receives the record "A" and inserts a cloned record "ClonedA" to cloud database with server generated id.
The record "ClonedA" has a special field "clientId" to keep track of the id of record "A"
In the next pulling, the record "ClonedA" is pulled to the client.
I use an adapter to do some transformation to the changes.
The adapter scans the changes and finds that the created record "ClonedA" has a field "clientId". Then it adds the "clientId" value(id of record "A") to the deleted changes.
The database will then insert the record "clonedA" and delete the local record "A", which acts like a replacement.

**Comment by heliojunior** [comment]
@KennanChan, do you have an example of the adapter you made?   I'm not finding an example adapter.

Thanks

**Comment by heliojunior** [comment]
Thanks @KennanChan, 

I implemented the same pattern and it worked great.

thanks for your help

**Comment by JeanGabriel109** [comment]
Thanks @KennanChan 

This helped me a lot, and to avoid having to create a localId column in every table on my backend, I also created an intermediate table. Whenever I’m updating or relating data in the backend during the push process, I go through a helper that checks whether the ID is already in numeric format. If it’s text, I look it up in this intermediate table, which allows me to resolve all related data.
